src/spell_check.c

```diff
@@ -0,0 +1,390 @@
+  } else {
+    // fieldMask has filtered all docs, this suggestions should not be returned
+    retVal = -1;
+  }
+  ReadIterator_Free(iter);
+
+end:
+  if (keyp) {
+    RedisModule_CloseKey(keyp);
+  }
+  return retVal;
+}
+
+static bool SpellCheck_IsTermExistsInTrie(Trie *t, const char *term, size_t len) {
+  rune *rstr = NULL;
+  t_len slen = 0;
+  float score = 0;
+  int dist = 0;
+  bool retVal = false;
+  TrieIterator *it = Trie_Iterate(t, term, len, 0, 0);
+  if (TrieIterator_Next(it, &rstr, &slen, NULL, &score, &dist)) {
+    retVal = true;
+  }
+  DFAFilter_Free(it->ctx);
+  free(it->ctx);
+  TrieIterator_Free(it);
+  return retVal;
+}
+
+static void SpellCheck_FindSuggestions(SpellCheckCtx *scCtx, Trie *t, const char *term, size_t len,
+                                       t_fieldMask fieldMask, RS_Suggestions *s) {
+  rune *rstr = NULL;
+  t_len slen = 0;
+  float score = 0;
+  int dist = 0;
+  size_t suggestionLen;
+
+  TrieIterator *it = Trie_Iterate(t, term, len, (int)scCtx->distance, 0);
+  while (TrieIterator_Next(it, &rstr, &slen, NULL, &score, &dist)) {
+    char *res = runesToStr(rstr, slen, &suggestionLen);
+    double score;
+    if ((score = SpellCheck_GetScore(scCtx, res, suggestionLen, fieldMask)) != -1) {
+      RS_SuggestionsAdd(s, res, suggestionLen, score);
+    } else {
+      free(res);
+    }
+  }
+  DFAFilter_Free(it->ctx);
+  free(it->ctx);
+  TrieIterator_Free(it);
+}
+
+static bool SpellCheck_ReplyTermSuggestions(SpellCheckCtx *scCtx, char *term, size_t len,
+                                            t_fieldMask fieldMask) {
+#define NO_SUGGESTIONS_REPLY "no spelling corrections found"
+#define TERM "TERM"
+
+  // searching the term on the term trie, if its there we just return false
+  // because there is no need to return suggestions on it.
+  if (SpellCheck_IsTermExistsInTrie(scCtx->sctx->spec->terms, term, len)) {
+    return false;
+  }
+
+  // searching the term on the exclude list, if its there we just return false
+  // because there is no need to return suggestions on it.
+  for (int i = 0; i < array_len(scCtx->excludeDict); ++i) {
+    RedisModuleKey *k = NULL;
+    Trie *t =
+        SpellCheck_OpenDict(scCtx->sctx->redisCtx, scCtx->excludeDict[i], REDISMODULE_READ, &k);
+    if (t == NULL) {
+      continue;
+    }
+    if (SpellCheck_IsTermExistsInTrie(t, term, len)) {
+      RedisModule_CloseKey(k);
+      return false;
+    }
+    RedisModule_CloseKey(k);
+  }
+
+  RS_Suggestions *s = RS_SuggestionsCreate();
+
+  RedisModule_ReplyWithArray(scCtx->sctx->redisCtx, 3);
+  RedisModule_ReplyWithStringBuffer(scCtx->sctx->redisCtx, TERM, strlen(TERM));
+  RedisModule_ReplyWithStringBuffer(scCtx->sctx->redisCtx, term, len);
+
+  SpellCheck_FindSuggestions(scCtx, scCtx->sctx->spec->terms, term, len, fieldMask, s);
+
+  // sorting results by score
+  qsort(s->suggestions, array_len(s->suggestions), sizeof(RS_Suggestion *),
+        (__compar_fn_t)RS_SuggestionCompare);
```

Comment:
this typename (__ compar_fn_t) doesn't exist on all compilers; it's just internal. Fix the function definition instead.